src/vec.rs

```diff
@@ -0,0 +1,364 @@
+        assert!(
+            max_value_size <= expected_value_size,
+            "max_value_size must be <= {}",
+            expected_value_size
+        );
+
+        Self {
+            memory,
+            max_value_size: header.max_value_size,
+            base: Address::from(0) + StableVecHeader::size(),
+            len: header.len,
+            _phantom: PhantomData,
+        }
+    }
+
+    fn save(&self) {
+        let header = StableVecHeader {
+            magic: *MAGIC,
+            version: LAYOUT_VERSION,
+            max_value_size: self.max_value_size,
+            fixed_value_size: T::FIXED_SIZE,
+            len: self.len,
+            _buffer: [0; 24],
+        };
+
+        write_struct(&header, Address::from(0), &self.memory);
+    }
+
+    pub fn clear(&mut self) {
+        self.len = 0;
+        self.save();
+    }
+
+    #[inline]
+    fn index_to_entry_addr(&self, index: usize) -> Address {
+        self.base + Bytes::from((Self::ENTRY_SIZE * index) as u64)
+    }
+
+    #[inline]
+    fn read_value(&self, index: usize) -> T {
+        let mut p = self.index_to_entry_addr(index);
+        // ret = read_struct(ptr, &self.memory);
+        let mut buf = if T::FIXED_SIZE {
+            vec![0; T::MAX_SIZE as usize]
+        } else {
+            let size = read_u32(&self.memory, p) as usize;
+            p += Bytes::from(4u64);
+            vec![0; size]
+        };
+        self.memory.read(p.get(), &mut buf);
+        T::from_bytes(buf)
+    }
+
+    #[inline]
+    fn write_value(&mut self, index: usize, value: &T) {
+        let value = value.to_bytes();
+        let mut p = self.index_to_entry_addr(index);
+        if !T::FIXED_SIZE {
+            write_u32(&self.memory, p, value.len() as u32);
+            p += Bytes::from(4u64);
+        }
+        self.memory.write(p.get(), &value);
+    }
+
+    pub fn insert(&mut self, index: usize, element: &T) {
+        let len = self.len();
+        assert!(index <= len);
+
+        // space for the new element
+        if len == self.capacity() {
+            self.reserve(1);
+        }
+
+        // Move all entries starting at index one to the right.
+        let src = self.index_to_entry_addr(index);
+        let dst = self.index_to_entry_addr(index + 1);
+        copy(
+            &self.memory,
+            src,
+            dst,
+            (len - index) as u64,
+            Self::ENTRY_SIZE,
+        );
+
+        self.write_value(index, element);
+
+        self.len += 1;
+        self.save();
+    }
+
+    pub fn remove(&mut self, index: usize) -> T {
+        let len = self.len();
+        assert!(index < len);
+
+        let ret = self.read_value(index);
+
+        // Shift everything down to fill in that spot.
+        copy::<M>(
+            &self.memory,
+            self.index_to_entry_addr(index + 1),
+            self.index_to_entry_addr(index),
+            (len - index - 1) as u64,
+            Self::ENTRY_SIZE,
+        );
```

Comment:
Same concern about shifting. Here, one potential solution would be to mark an element as deleted, and whenever the user provides an `index`, we modify the `index` to go to the appropriate location based on which elements have been deleted.

Reply:
Actually, that is exactly what we need for the binary heap.

Reply:
But we still need need instructions to access (read/overwrite) elements within the vector. So i guess it would still not quite be a stack.

Reply:
Fair enough. A growable array with `push` and `pop` does seem generic enough, and I can use-cases for this even without the `insert` and `remove`. So, yes, I'd suggest removing these APIs.